src/python/pants/engine/legacy/graph.py

```diff
@@ -245,8 +255,7 @@ def _inject_address_specs(self, address_specs: AddressSpecs):
 
     logger.debug('Injecting address specs to %s: %s', self, address_specs)
     with self._resolve_context():
-      thts, = self._scheduler.product_request(TransitiveHydratedTargets,
-                                              [address_specs])
+      thts, = self._scheduler.product_request(TransitiveHydratedTargets, [address_specs])
```

Comment:
This line is the root of the problem. Once we change `HydratedTarget` and `TargetAdaptor` to use `Address` instead of `BuildFIleAddress`, this code path will no longer be able to get the `BuildFileAddress`es needed by V1 for backwards compatibility.

Two possible approaches:
1) Use the new path `Address -> BuildFileAddress` to get this working ad hoc by proactively converting each target in `TransitiveHydratedTargets`.
2) Introduce `LegacyHydratedTarget` and `LegacyTransitiveHydratedTargets`, which still use `BuildFileAddress`. 

I think #2 is the way to go. #1 would be difficult to get right and would require lots of MyPy ignores.